### ISSUE TITLE
Issue for the ODE function 

### ISSUE THREAD

**Issue by harryjo97** [DeepGraphLearning/ContinuousGNN#2]
If I understood the paper and the code correctly, 
I believe this line of code for the ODE function : 

https://github.com/DeepGraphLearning/ContinuousGNN/blob/dd61619f3727ad83ae6580c33c016024d9f5d190/src/gnn.py#L46

should be changed to 
`f = alph * 0.5 * ax - x + self.x0`

please let me know if I am wrong
Thank you.

**Comment by KimMeen** [comment]
Same question here. 
Could you help let us know if this is correct? @lpxhonneux

**Comment by sophie-xhonneux** [comment]
The original line is correct, we had self-loops during pre-processing and then the equations work out.

Edit: This is a confusing answer, see my later reply for a clearer explanation.

**Comment by harryjo97** [comment]
I guess the self-loops are added by the following line :
https://github.com/DeepGraphLearning/ContinuousGNN/blob/dd61619f3727ad83ae6580c33c016024d9f5d190/src/train.py#L42

I think the variables in `class ODEFunc(nn.Module)` contains the following values when used in `train.py` line 75 :

`f` = dH(t)/dt
`x` = H(t)
`ax` = (I + D<sup>-1/2</sup> Adj D<sup>-1/2</sup>)H(t)
`self.x0` = H(0)

Thus, I believe that the equation (7) of the original paper should be constructed as :
`f = alph * 0.5 * ax - x + self.x0`

**Comment by KimMeen** [comment]
@lpxhonneux Thank you for your timely reply, could you explain a little bit about what's going wrong on @harryjo97's derivation above?

**Comment by sophie-xhonneux** [comment]
During pre-processing we turn the adjacency matrix into the normalised Adjacency L, hence the variable ax refers to the matrix multiplication of the normalised Adjacency L (note that A in the paper refers to 0.5*(I+L)) times the matrix of node features X. The ODE requires us to compute (A-I)X, thus yielding 0.5*(ax-x).

**Comment by KimMeen** [comment]
Thank you @lpxhonneux. When ```alpha != 1```, should it be ```f=((alpha-2) * 0.5)x + alpha * 0.5 * ax + x_0```, where ```A = alpha * 0.5 * (I + L)``` and ```ax = L * X```? Your case makes sense to me if ```alpha=1```.
